consensus/ninja_tortoise.go

```diff
@@ -0,0 +1,444 @@
+	stack := list.New()
+	for b, _ := range blocks {
+		stack.PushFront(b)
+	}
+	layerCounter := make(map[mesh.LayerID]int)
+	set := make(map[mesh.BlockID]struct{})
+	for b := stack.Front(); b != nil; b = stack.Front() {
+		a := stack.Remove(stack.Front()).(mesh.BlockID)
+		block, found := blockCache[a]
+		if !found {
+			panic(fmt.Sprintf("error block not found ID %d", block.ID()))
+		}
+		layerCounter[block.Layer()]++
+		foo(block)
+		//push children to bfs queue
+		for _, bChild := range block.ViewEdges {
+			if blockCache[bChild].Layer() >= layer { //dont traverse too deep
+				if _, found := set[bChild]; !found {
+					set[bChild] = struct{}{}
+					stack.PushFront(bChild)
+				}
+			}
+		}
+	}
+	return layerCounter
+}
+
+func globalOpinion(v vec, layerSize uint32, delta float64) vec {
+	threshold := float64(GlobalThreshold*delta) * float64(layerSize)
+	if float64(v[0]) > threshold {
+		return Support
+	} else if float64(v[1]) > threshold {
+		return Against
+	} else {
+		return Abstain
+	}
+}
+
+func (ni *ninjaTortoise) updateCorrectionVectors(p votingPattern, bottomOfWindow mesh.LayerID) {
+	foo := func(x *mesh.Block) {
+		for _, bid := range ni.tEffectiveToBlocks[p] { //for all b who's effective vote is p
+			b := ni.blocks[bid]
+			if _, found := ni.tExplicit[b.Id][x.Layer()]; found { //if Texplicit[b][x]!=0 check correctness of x.layer and found
+				ni.Debug(" blocks pattern %d block %d layer %d", p, b.ID(), b.Layer())
+				if _, found := ni.tCorrect[b.Id]; !found {
+					ni.tCorrect[b.Id] = make(map[mesh.BlockID]vec)
+				}
+				vo := ni.tVote[p][x.ID()]
+				ni.Debug("vote from pattern %d to block %d layer %d vote %d ", p, x.ID(), x.Layer(), vo)
+				ni.tCorrect[b.Id][x.ID()] = vo.Negate() //Tcorrect[b][x] = -Tvote[p][x]
+				ni.Debug("update correction vector for block %d layer %d , pattern %d vote %d for block %d ", b.ID(), b.Layer(), p, ni.tCorrect[b.Id][x.ID()], x.ID())
+			} else {
+				ni.Debug("block %d from layer %d dose'nt explicitly vote for layer %d", b.ID(), b.Layer(), x.Layer())
+			}
+		}
+	}
+
+	forBlockInView(ni.tPattern[p], ni.blocks, bottomOfWindow, foo)
+}
+
+func (ni *ninjaTortoise) updatePatternTally(newMinGood votingPattern, botomOfWindow mesh.LayerID) {
+	ni.Debug("update tally pbase id:%d layer:%d p id:%d layer:%d", ni.pBase.id, ni.pBase.Layer(), newMinGood.id, newMinGood.Layer())
+
+	correctionMap := make(map[mesh.BlockID]vec)
+	effCountMap := make(map[mesh.LayerID]int)
+	foo := func(b *mesh.Block) {
+		if eff, found := ni.tEffective[b.ID()]; found {
+			effCountMap[eff.Layer()] = effCountMap[eff.Layer()] + 1
+			if p, found := ni.tGood[eff.Layer()]; found && eff == p {
+				for k, v := range ni.tCorrect[b.ID()] {
+					correctionMap[k] = correctionMap[k].Add(v)
+				}
+			}
+		}
+	}
+
+	forBlockInView(ni.tPattern[newMinGood], ni.blocks, ni.pBase.Layer()+1, foo) //ni.pBase.LayerID + 1 because we already copied pbase's votes
+
+	for idx := ni.pBase.Layer(); idx < newMinGood.Layer(); idx++ {
```

Comment:
I think that this for is wrong for several reasons -
1. you might have relevant good effective patterns before pbase
2. you might have good layers after pbase that are irrelevant for this count (because they aren't good effective for any block in p's view)

IMO instead you should iterate on the key's of effCountMap